Documentation/network/lb-ipam.rst

```diff
@@ -0,0 +1,382 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    https://docs.cilium.io
+
+.. _lb_ipam:
+
+********************************************
+LoadBalancer IP Address Management (LB IPAM)
+********************************************
+
+LB IPAM is a feature that allows Cilium to assign IP addresses to Services of
+type ``LoadBalancer``. This functionally is usually left up to a cloud provider,
+however, when deploying in a private cloud environment, these facilities are not
+always available.
+
+LB IPAM work in conjunction with features like the :ref:`bgp_control_plane`. Where
+LB IPAM is responsible for allocation and assigning of IPs to Service objects and
+other features are responsible for load balancing and/or advertisement of these
+IPs. 
+
+LB IPAM is implicitly enabled when dependant features are enabled.
+
+Pools
+#####
+
+LB IPAM has the notion of IP Pools which the administrator can create to tell 
+Cilium which IP ranges can be used to allocate IPs from.
+
+A basic IP Pools with both an IPv4 and IPv6 range looks like.
+
+.. code-block:: yaml
+
+    apiVersion: "cilium.io/v2alpha1"
+    kind: CiliumLoadBalancerIPPool
+    metadata:
+      name: "blue-pool"
+    spec:
+      cidrs:
+      - cidr: "10.0.10.0/24"
+      - cidr: "2004::0/64"
+
+After adding the pool to the cluster, it appears like so.
+
+.. code-block:: shell-session
+
+    $ kubectl get ippools                           
+    NAME        DISABLED   CONFLICTING   IPS AVAILABLE   AGE
+    blue-pool   false      false         65788           2s
+
+.. note::
+    The amount of available IPs in the pool is less than the actual sum of all
+    usable IPs in the CIDRs because the allocation logic is limited to 65536 IPs
+    per CIDR. CIDRs containing more than 65536 IPs can be broken up into multiple
+    smaller CIDRs to achieve full utilization.
+
+Service Selectors
+-----------------
+
+IP Pools have an optional ``.spec.serviceSelector`` field which allows administrators
```

Comment:
This is a good idea -- an alternative would be a complicated RBAC setup with a WebHook -- but using the name Selector feels a bit "backwards" to me.

My mental model is that you label a resource, and consume that resource with a selector. In other words, Nodes are labeled and Pods have LabelSelectors. Likewise, Pods have labels and Services have LabelSelectors. In this vein, the IPPool is the resource, and the Service should have a selector.

Obviously this doesn't work, since we're not going to update the `v1.Service` object :-).

So, can we rename this construct? Something to imply this is a restriction?

Reply:
Used this naming convention since we use `podSelector` and `nodeSelector` in other places to select pods and nodes. So it seems logical to name this a `serviceSelector` to match services. I am not sure which restriction we have by doing it like this, can you elaborate?